src/testdrive/src/action/s3.rs

```diff
@@ -93,3 +100,172 @@ impl Action for PutObjectAction {
             .map_err(|e| format!("putting s3 object: {}", e))
     }
 }
+
+pub struct AddBucketNotifications {
+    bucket: String,
+    events: Vec<String>,
+
+    queue: String,
+
+    bucket_prefix: String,
+}
+
+pub fn build_add_notifications(mut cmd: BuiltinCommand) -> Result<AddBucketNotifications, String> {
+    let events = cmd
+        .args
+        .opt_string("events")
+        .map(|a| a.split(",").map(|s| s.to_string()).collect())
+        .unwrap_or_else(|| vec!["s3:ObjectCreated:*".to_string()]);
+
+    let bucket_prefix = cmd
+        .args
+        .opt_string("bucket_prefix")
+        .unwrap_or_else(|| "materialize-ci-*".into());
+
+    Ok(AddBucketNotifications {
+        bucket: cmd.args.string("bucket")?,
+        events,
+        queue: cmd.args.string("queue")?,
+        bucket_prefix,
+    })
+}
+
+#[async_trait]
+impl Action for AddBucketNotifications {
+    async fn undo(&self, _state: &mut State) -> Result<(), String> {
+        Ok(())
+    }
+
+    async fn redo(&self, state: &mut State) -> Result<(), String> {
+        let result = state
+            .sqs_client
+            .create_queue(CreateQueueRequest {
+                queue_name: self.queue.clone(),
+                ..Default::default()
+            })
+            .await;
+
+        // get queue properties used for the rest of the mutations
+
+        let queue_url = match result {
+            Ok(r) => r
+                .queue_url
+                .expect("queue creation should always return the url"),
+            Err(RusotoError::Service(CreateQueueError::QueueNameExists(q))) => {
+                let resp = state
+                    .sqs_client
+                    .get_queue_url(GetQueueUrlRequest {
+                        queue_name: q,
+                        queue_owner_aws_account_id: None,
+                    })
+                    .await
+                    .map_err(|e| {
+                        format!(
+                            "when trying to get sqs queue url for already-existing queue: {}",
+                            e
+                        )
+                    })?;
+                resp.queue_url
+                    .expect("successfully getting the url gets the url")
+            }
+            Err(e) => return Err(e.to_string()),
+        };
+
+        let queue_arn = state
+            .sqs_client
+            .get_queue_attributes(GetQueueAttributesRequest {
+                attribute_names: Some(vec!["QueueArn".to_string()]),
+                queue_url: queue_url.clone(),
+            })
+            .await
+            .map_err(|e| format!("getting queue {} attributes: {}", self.queue, e))?
+            .attributes
+            .expect("the result should not be empty")
+            .remove("QueueArn")
+            .unwrap();
```

Comment:
yeah, I believe (just a hunch) that they are present just because the AWS API schemas don't provide enough information to ever mark something as non-optional. I'll switch to errors though.